libs/mock-block-dock/src/datastore/traverse.ts

```diff
@@ -1,284 +1,539 @@
+  isEntityVertex,
+  KnowledgeGraphOutwardEdge,
+  OntologyRootedEdges,
   OutgoingLinkEdge,
+  OutwardEdge,
   Subgraph,
+  SubgraphRootType,
+  TimeInterval,
+  Vertex,
 } from "@blockprotocol/graph";
 import { addKnowledgeGraphEdgeToSubgraphByMutation } from "@blockprotocol/graph/internal";
 import {
   getIncomingLinksForEntity,
   getLeftEntityForLinkEntity,
   getOutgoingLinksForEntity,
+  getRightEntityForLinkEntity,
+  intervalIntersectionWithInterval,
+  mapElementsIntoRevisions,
 } from "@blockprotocol/graph/stdlib";
-import { getRightEntityForLinkEntity } from "@blockprotocol/graph/stdlib-temporal";
 
-import { typedEntries } from "../util";
-import { ResolveMap } from "./traverse/resolve-map";
-import { PartialDepths, TraversalContext } from "./traverse/traversal-context";
+import { get, mustBeDefined, typedEntries } from "../util";
 
-/** @todo - doc */
-/** @todo - Update this to take an interval instead of always being "latest" */
-/** @todo - Update this to handle ontology edges */
-/** @todo - Update this to use temporal versioning information */
-export const traverseElement = (
-  traversalSubgraph: Subgraph,
-  elementIdentifier: GraphElementVertexId,
-  datastore: Subgraph,
-  traversalContext: TraversalContext,
-  currentTraversalDepths: PartialDepths,
-) => {
-  const unresolvedDepths = traversalContext.insert(
-    elementIdentifier,
-    currentTraversalDepths,
+const TIMESTAMP_PLACEHOLDER = `TIMESTAMP_PLACEHOLDER` as const;
+
+/**
+ * Advanced type to recursively search a type for `EntityValidInterval` and patch those occurrences by removing the
+ * definition of the "validInterval" property.
+ */
+type PatchEntityValidInterval<ToPatch extends unknown> = ToPatch extends object
+  ? ToPatch extends EntityValidInterval
+    ? Omit<ToPatch, "validInterval">
+    : { [key in keyof ToPatch]: PatchEntityValidInterval<ToPatch[key]> }
+  : ToPatch;
+
+/**
+ * A patched {@link Subgraph} type which is partially incomplete. Entity-to-entity edges aren't fully specified, as
+ * to avoid many intermediary rewrites, the timestamps of the edges are filled in as a post-processing step and are left
+ * unspecified until then.
+ */
+export type TraversalSubgraph<
+  Temporal extends boolean,
+  RootType extends SubgraphRootType<Temporal> = SubgraphRootType<Temporal>,
+> = Omit<Subgraph<Temporal, RootType>, "edges"> & {
+  edges: PatchEntityValidInterval<
+    OntologyRootedEdges & {
+      [entityId: EntityId]: Record<
+        typeof TIMESTAMP_PLACEHOLDER,
+        KnowledgeGraphOutwardEdge[]
+      >;
+    }
+  >;
+};
+
+export const isTemporalTraversalSubgraph = <
+  RootType extends SubgraphRootType<boolean> = SubgraphRootType<boolean>,
+>(
+  traversalSubgraph: TraversalSubgraph<boolean, RootType>,
+): traversalSubgraph is TraversalSubgraph<true, RootType> => {
+  // this cast is safe as we're only checking against undefined
+  return (
+    (traversalSubgraph as TraversalSubgraph<true>).temporalAxes !== undefined
```

Comment:
I don't think this will avoid a crash, because if the assertion is incorrect (and the field does not exist) then it will be equivalent to undefined `({} as any).foo === undefined`